src/main/java/net/minecraftforge/client/gui/config/ControlCreator.java

```diff
@@ -0,0 +1,519 @@
+            this.label = title;
+        }
+
+        public void addValidator(Predicate<T> newValidator) {
+            validator = validator.and(newValidator);
+        }
+
+        public boolean isValid(T newValue) {
+            return validator.test(newValue);
+        }
+
+        public void addSaver(Consumer<T> newSaver) {
+            saver = saver.andThen(newSaver);
+        }
+
+        /**
+         * Saves a valid new value.
+         */
+        public void save(T newValue) {
+            saver.accept(newValue);
+        }
+
+        public void addUpdateResponder(BiConsumer<Boolean, T> newUpdateResponder) {
+            updateResponder = updateResponder.andThen(newUpdateResponder);
+        }
+
+        public void onUpdate(T newValue) {
+            onUpdate(isValid(newValue), newValue);
+        }
+
+        /**
+         * Responds to value changes.
+         * Called to set up the Interactor after initialisation.
+         * Called when the user interacts with the Widget directly and changes the value.
+         * Called when the user clicks the undo/reset buttons.
+         */
+        public void onUpdate(boolean isValid, T newValue) {
+            updateResponder.accept(isValid, newValue);
+        }
+
+        public <D> void addData(DataKey<D> key, @Nullable D value) {
+            extraData.put(key, value);
+        }
+
+        @Nullable
+        public <D> D getData(DataKey<D> key) {
+            return (D) extraData.get(key);
+        }
+
+        /**
+         * For making a system similar to a type safe map and the capabilities system.
+         * Lets us store arbitrary extra data in a strongly typed manner.
+         */
+        static class DataKey<T> {
+        }
+
+    }
+
+    public static ControlCreator getDefaultCreator() {
+        return DefaultHolder.INSTANCE;
+    }
+
+    public <T> void createAndInitialiseInteractionWidget(Interactor<T> interactor) {
+        createInteractionWidget(interactor);
+        interactor.onUpdate(interactor.initialValue);
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    public <T> void createInteractionWidget(Interactor<T> interactor) {
+        T value = interactor.initialValue;
+        if (value instanceof Boolean)
+            createBooleanInteractionWidget((Interactor<Boolean>) interactor);
+        else if (value instanceof Number)
+            createNumericInteractionWidget((Interactor<? extends Number>) interactor);
+        else if (value instanceof Enum<?>)
+            createEnumInteractionWidget((Interactor<Enum>) interactor);
+        else if (value instanceof String)
+            createStringInteractionWidget((Interactor<String>) interactor);
+        else if (value instanceof List<?>)
+            createListInteractionWidget((Interactor<List<?>>) interactor);
+        else
+            throw new IllegalStateException("Don't know how to make a widget for " + value);
```

Comment:
How likely is this going to happen? Can this happen with a mod using custom objects?

Reply:
Unlikely.
I've covered all the common values (booleans, numbers, strings, enums) + lists. I haven't yet implemented `Config`s + if something similar to #6464 is merged this (specifically `createNumericInteractionWidget`) will need to be extended to support the new types.
However, any `Exception`s thrown by this method are caught in the `tryCreateInteractor` methods in `CategoryConfigScreen` and `ListConfigScreen`.

Mod's using custom objects may also trigger this, but any exceptions will be caught and a simple, uneditable, `toString()` representation will be displayed instead of a widget with a tooltip telling the user to edit the value in the config file.